snmp/datadog_checks/snmp/models.py

```diff
@@ -21,23 +22,72 @@ class OID(object):
 
     def __init__(self, value):
         # type: (Union[Sequence[int], str, ObjectName, ObjectIdentity, ObjectType]) -> None
+        parts = None  # type: Optional[Tuple[int, ...]]
+
         try:
             parts = parse_as_oid_tuple(value)
         except CouldNotDecodeOID:
-            raise  # Explicitly re-raise this exception.
+            raise  # Invalid input.
+        except UnresolvedOID:
+            if isinstance(value, ObjectType):
+                # An unresolved `ObjectType(ObjectIdentity('<MIB>', '<symbol>'))`.
+                parts = None
+            elif isinstance(value, ObjectIdentity):
+                # An unresolved `ObjectIdentity('<MIB>', '<symbol>')`.
+                parts = None
+            else:  # pragma: no cover
+                raise RuntimeError('Unexpectedly treated {!r} as an unresolved OID'.format(value))
 
-        # Let's make extra sure we didn't mess up.
-        if not isinstance(parts, tuple):
-            raise RuntimeError(
-                'Expected result {!r} of parsing value {!r} to be a tuple, but got {}'.format(parts, value, type(parts))
-            )  # pragma: no cover
+        # Resolve the `ObjectIdentity` which we can use to resolve the MIB name of the OID (for metric naming).
+        # PySNMP objects may contain MIB information already, so check for them in priority.
+        if isinstance(value, ObjectType):
+            # No other choice than to use private API here.
+            object_identity = value._ObjectType__args[0]
+            if not isinstance(object_identity, ObjectIdentity):  # pragma: no cover
+                raise RuntimeError('Expected {!r} to be an `ObjectIdentity` instance'.format(object_identity))
+        elif isinstance(value, ObjectIdentity):
+            object_identity = value
+        else:
+            # Fallback.
+            if parts is None:  # pragma: no cover
+                raise RuntimeError('`parts` should have been set')
+            object_identity = ObjectIdentity(parts)
```

Comment:
Given current usage, we should change `parse_metrics` in the same patch, otherwise we create twice as many ObjectIdentity as necessary.

Reply:
Yes, currently we'll be creating unnecessary `ObjectIdentity` objects for most metrics (as they're given in `x.y.z.t` format).

I'll push a commit. 😄 Diff will grow a bit because I'll also have to update the unit tests to get rid of the spy object and inspect the lists of `OID` directly (which is actually nicer).